ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/HD/BackingStore/Mock.hs

```diff
@@ -0,0 +1,273 @@
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE NamedFieldPuns             #-}
+
+module Test.Ouroboros.Storage.LedgerDB.HD.BackingStore.Mock (
+    -- * Types
+    Err (..)
+  , ID (..)
+  , Mock (..)
+  , MockValueHandle (..)
+  , emptyMock
+    -- * Type classes
+  , ApplyDiff (..)
+  , DiffSize (..)
+  , EmptyValues (..)
+  , HasOps
+  , KeysSize (..)
+  , LookupKeys (..)
+  , LookupKeysRange (..)
+  , MakeDiff (..)
+  , ValuesLength (..)
+    -- * State monad to run the mock in
+  , MockState (..)
+  , runMockState
+    -- * Mocked @'BackingStore'@ operations
+  , mBSClose
+  , mBSCopy
+  , mBSVHClose
+  , mBSVHRangeRead
+  , mBSVHRead
+  , mBSValueHandle
+  , mBSWrite
+  , mGuardBSClosed
+  , mGuardBSVHClosed
+  ) where
+
+import           Control.Monad
+import           Control.Monad.Except (ExceptT (..), MonadError (throwError),
+                     runExceptT)
+import           Control.Monad.State (MonadState, State, StateT (StateT), gets,
+                     modify, runState)
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+
+import           Ouroboros.Consensus.Block.Abstract (SlotNo, WithOrigin (..))
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD.BackingStore as BS
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+data Mock vs = Mock {
+    backingValues :: vs
+  , backingSeqNo  :: WithOrigin SlotNo
+  , copies        :: Set BS.BackingStorePath
+  , isClosed      :: Bool
+    -- | Track whether value handles have been closed.
+  , valueHandles  :: Map ID Bool
+    -- | The next id to use if a new value handle is opened.
+  , nextId        :: ID
+  }
+  deriving stock (Show, Eq)
+
+data MockValueHandle values = MockValueHandle {
+    getId  :: ID
+  , values :: values
+  , seqNo  :: WithOrigin SlotNo
+  }
+  deriving stock Show
+
+instance Eq (MockValueHandle vs) where
+  x == y = getId x == getId y
+
+instance Ord (MockValueHandle vs) where
+  x <= y = getId x < getId y
+
+-- | An ID for a mocked value handle.
+newtype ID = ID Word
+  deriving stock (Show, Eq, Ord)
+  deriving newtype Num
+
+-- | An empty mock state.
+emptyMock :: EmptyValues vs => Mock vs
+emptyMock = Mock {
+    backingValues = emptyValues
+  , backingSeqNo  = Origin
+  , copies        = Set.empty
+  , isClosed      = False
```

Comment:
Now that I see this starts as `False` I see that we start with an opened backing store. Would it make sense to start with a closed one and then open it as a command? Something like `BSOpen` which opens a BS? I imagine that there might be corner cases that await there (double opening, open-close-open, etc)

Reply:
I've been experimenting with this but I'm working myself into a corner -- I think it's because a `BackingStore` is a datatype: opening a `BackingStore` creates a new datatype, closing a `BackingStore` is an operation of the datatype. I'll have to think about this more

Reply:
There is now an action that reopens a backing store either from a file or from provided values, but only if the backing store is closed, which is guaranteed by the precondition. This mimicks the way the ChainDB state-machine tests are set up for reopening ChainDBs.